src/agent/block_store_services/block_store_client.js

```diff
@@ -43,223 +60,354 @@ class BlockStoreClient {
+                    block_md: store_block_md,
+                };
+            } catch (err) {
+                dbg.error('S3 operation failed:', err, bs_info && _.omit(bs_info, 'access_keys'));
+                if (err.code === 'NoSuchBucket') {
+                    block_store_info_cache.invalidate_key(options.address);
+                    throw new RpcError('STORAGE_NOT_EXIST', `s3 bucket ${bs_info[options.address].target_bucket} not found. got error ${err}`);
+                } else if (err.code === 'AccessDenied') {
+                    block_store_info_cache.invalidate_key(options.address);
+                    throw new RpcError('AUTH_FAILED', `access denied to the s3 bucket ${bs_info[options.address].target_bucket}. got error ${err}`);
+                }
+                throw new Error(err.message || 'unknown error');
+            }
+        }, timeout);
 
-                        } else {
-                            // parse the error and throw Error object
-                            await this._throw_s3_err(res);
-                        }
-                    }
+    }
 
-                } catch (err) {
-                    dbg.error('encountered error on _delegate_read_block_s3:', err);
-                    const error = _.pick(err, 'message', 'code');
-                    return rpc_client.block_store.handle_delegator_error({
-                        error,
-                        usage: {
-                            count: 1,
-                            size: params.block_md.size,
-                        },
-                        op_type: 'READ'
-                    }, options);
-                }
-            })
-            .timeout(timeout);
+    _update_usage_stats(rpc_client, usage, address, read_or_write) {
+        if (!this.io_stats) {
+            this.io_stats = {};
+        }
+        if (!this.io_stats[address]) {
+            this.io_stats[address] = {
+                read_count: 0,
+                read_bytes: 0,
+                write_count: 0,
+                write_bytes: 0,
+                rpc_client
+            };
+        }
+        if (read_or_write === 'WRITE') {
+            this.io_stats[address].write_count += usage.count;
+            this.io_stats[address].write_bytes += usage.size;
+        } else if (read_or_write === 'READ') {
+            this.io_stats[address].read_count += usage.count;
+            this.io_stats[address].read_bytes += usage.size;
+        } else {
+            throw new Error('_update_usage_stats should except READ or WRITE only');
+        }
+    }
 
+    async send_usage_stats() {
+        for (const address in this.io_stats) {
+            if (this.io_stats[address]) {
+                const rpc_client = this.io_stats[address].rpc_client;
+                const io_stats = _.omit(this.io_stats[address], 'rpc_client');
+                this.io_stats[address] = null;
+                await rpc_client.block_store.update_store_usage(io_stats, { address });
+            }
```

Comment:
Missing try catch. 
Also need to set to null after successful send. 
Also better delete from Map instead of setting to null, so better use Map instead of plain object.
Also since this iteration is doing async calls it is best to first loop through and prepare a list of updates and then send them in Promise.all.
I would consider rewriting this function.